python/cudf/cudf/core/join/join.py

```diff
@@ -234,14 +234,57 @@ def _compute_join_keys(self):
                         for on in _coerce_to_tuple(self.right_on)
                     ]
                 )
+        elif self.on:
+            on_names = _coerce_to_tuple(self.on)
+            for on in on_names:
+                # If `on` if provided, checks whether merging on
+                # columns, indexes or merging  index with column
+                if (
+                    on in self.lhs._data.keys()
+                    and on not in self.rhs._data.keys()
+                ):
+                    # case1: merge on lhs column with rhs index
+                    if on in self.rhs.index.names:
+                        left_keys.extend([_Indexer(name=on, column=True)])
+                        right_keys.extend([_Indexer(name=on, index=True)])
```

Comment:
Should raise if `on` is not in `rhs.index.names`?

Reply:
`_Indexer()` will raise a `KeyError` if this is the case.

Reply:
Ditto below. Looping over `on` (or `left_on` and right_on`) for doing validation will not only add overhead but also [It's Easier To Ask For Forgiveness Than Permission](https://docs.python.org/3/glossary.html#EAFP).

Reply:
> _Indexer() will raise a KeyError if this is the case.

Will it? It seems like `on` will just get silently dropped? E.g.:
```
lhs: {'X': [1, 2, 3], 'b':[4, 5, 6]} index:{'L0': [10, 11, 12]}
rhs: {'a': [4, 5, 6], 'b':[1, 2, 3]} index:{'L0': [10, 11, 12]}
```
`on == 'X'`, technically it's not valid but here it just falls through and not get picked up by any subsequent code (unless that's the desired behavior).